### ISSUE TITLE
Add decl bindings for simple let statements [phab:D958]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1282]
_Created by **djrenren** at 2016-07-11 21:37:03_: 

This commit adds the createion of variable nodes for local let
bindings

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-11 21:40:29_:

  - Added a comment and renamed local_decl_vname

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-11 21:52:49_:

How are `crate_id`s generated?

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-07-11 23:07:05_:

That's an excellent question... it's probably generated based on order :/ so I can totally just do a lookup and get the crate name. but honestly the corpus should probably be the difference there, not the sig... will post a revision. local variables will always have a crate_id that is equal to zero. So for now we can ignore it but I'll leave a note. This would make the vname schema for variables look like:

```
{
    "signature": "variable:123456"
    "corpus": "somecrate"
    "language": "rust"
}
```

does that sound good?

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-11 23:18:04_:

  - Remove CrateNum from sig for variables

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-11 23:58:37_:

nit: you indented this in the other test; choose one style for consistency.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-11 23:58:37_:

Is there a way to do this that doesn't involve string comparison? Do the HIR objects you have available to you have some flag that shows whether they are implicit code?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-11 23:58:37_:

Is there prominent documentation somewhere that says that bad things will happen if crates are not given globally unique corpora? I think this is one of your core assumptions about the way the indexer is invoked.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-07-12 00:21:38_:

Well I'd written out a really long explanation of why we can't but then I found a very counterintuitive but correct way of doing this [from clippy](https://github.com/Manishearth/rust-clippy/blob/ad1cd990549fdfc8ae9dcd4ea7eea851017eb042/clippy_lints/src/utils/higher.rs#L121). On it.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-07-12 00:21:38_:

So currently, yes. In the future I'd like to ingest some sort of config file. that aides in the construction of `VName`s. `Corpus` has really evolved into a `VName` generator rather than a node emitter (as I originally had it). So it would make sense to merge that logic in there. So I could put up a big scary warning but for now, since I'm not doing anything cross-crate nothing is broken per-se and it's very possible that that restriction isn't a hard one.

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-12 00:42:09_:

  - Check for desugaring w/o string comparison

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-12 00:59:53_:

In a future diff you might consider indexing implicit declarations; LGTM (but please make sure to document the corpus precondition).

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-07-12 00:59:53_

**Comment by kythe** [comment]
_**djrenren** closed via commit 1bb8857bf1522ea6d6f1be638c2f319b84700f8e at 2016-07-12 01:01:17_
